src/olympia/amo/middleware.py

```diff
@@ -239,7 +239,9 @@ def process_request(self, request):
             if writable_method:
                 return JsonResponse({'error': self.ERROR_MSG}, status=503)
         elif request.method == 'POST':
-            return render(request, 'amo/read-only.html', status=503)
+            response = render(request, 'amo/read-only.html', status=503)
+            response.render()  # Force render to return an HttpResponse
+            return response
```

Comment:
this reads weird - what's going on with the read-only case that we have to force `.render()` on the response before returning?

Reply:
Middlewares that return a response instead of letting the request be processed by the view need to directly return an `HttpResponse`. If you return a `TemplateResponse` without rendering it, it will never be rendered, because that normally happens in view handling code that is being bypassed.

Reply:
Where is the view handling code that is being bypassed that would ordinarily do `render()`?  django code?

(I sometimes think we should rewrite all our old middlewares to use modern style callables rather than using `MiddlewareMixin`)

Reply:
It's in django yes. 

This is alluded to in the docs, it might be surprising at first glance but makes sense given what happens behind the scenes:
> (...) the `get_response` callable for the last middleware in the chain won’t be the actual view but rather a wrapper method from the handler which takes care of applying view middleware, calling the view with appropriate URL arguments, and applying template-response and exception middleware.
> (...)
> If one of the layers decides to short-circuit and return a response without ever calling its get_response, none of the layers of the onion inside that layer (including the view) will see the request or the response. The response will only return through the same layers that the request passed in through.

Returning a response at this step is allowed, but [it bypasses `get_response()` entirely](https://github.com/django/django/blob/99532fdadf0b2bd065b1f50b4ec14e1b0e53f8a3/django/utils/deprecation.py#L117), and that function is [what does the rendering](https://github.com/django/django/blob/99532fdadf0b2bd065b1f50b4ec14e1b0e53f8a3/django/core/handlers/base.py#L204).

If we don't call `render()` ourselves, all middlewares that we went through that need to do something with the response content (`CommonMiddleware`, `GZipMiddleware`) will fail as soon as they try to access it as rendering should have taken place by now, but didn't because we bypassed it. If we remove them, it will fail at the very end, when the test is accessing the response content, for the same reason.

Reply:
I'm open to suggestions to reword that comment since evidently it's not very clear as it stands now :)

Reply:
it's unfortunate the function is named `render` when it doesn't actually render... but something like "render the lazy response into a HttpResponse"?  I'm assuming that the response **always** has a `render` method too?

Reply:
That's a good point. Maybe I could get rid of that `render()` function altogether, and use `TemplateResponse` directly everywhere ? We'd still need to call `response.render()` here, but it would be less confusing.

Reply:
https://github.com/mozilla/addons-server/pull/18295/commits/20b612358178e8fb6d5a55212327c8c69f06e52b